sqlmesh/core/context.py

```diff
@@ -450,22 +452,80 @@ def run(
             True if the run was successful, False otherwise.
         """
         environment = environment or self.config.default_target_environment
+
+        if not skip_janitor and environment.lower() == c.PROD:
+            self._run_janitor()
+
         self.notification_target_manager.notify(
             NotificationEvent.RUN_START, environment=environment
         )
-        try:
-            success = self.scheduler(environment=environment).run(
-                environment,
-                start=start,
-                end=end,
-                execution_time=execution_time,
-                ignore_cron=ignore_cron,
-            )
-        except Exception as e:
-            self.notification_target_manager.notify(
-                NotificationEvent.RUN_FAILURE, traceback.format_exc()
+
+        env_check_attempts_num = max(
+            1,
+            self.config.run.environment_check_max_wait
+            // self.config.run.environment_check_interval,
+        )
+
+        def _block_until_finalized() -> str:
+            for _ in range(env_check_attempts_num):
+                assert environment is not None
+                environment_state = self.state_sync.get_environment(environment)
+                if not environment_state:
+                    raise SQLMeshError(f"Environment '{environment}' was not found.")
+                if environment_state.finalized_ts:
+                    return environment_state.plan_id
+                logger.warning(
+                    "Environment '%s' is being updated by plan '%s'. Retrying in %s seconds...",
+                    environment,
+                    environment_state.plan_id,
+                    self.config.run.environment_check_interval,
+                )
+                time.sleep(self.config.run.environment_check_interval)
+            raise SQLMeshError(
+                f"Exceeded the maximum wait time for environment '{environment}' to become ready."
```

Comment:
link to the documentation on how to change these settings and what it means when this error is hit

Reply:
improved the error message